near-sdk/src/collections/vector.rs

```diff
@@ -13,6 +15,7 @@ const ERR_INDEX_OUT_OF_BOUNDS: &[u8] = b"Index out of bounds";
 /// An iterable implementation of vector that stores its content on the trie.
 /// Uses the following map: index -> element.
 #[derive(BorshSerialize, BorshDeserialize)]
+#[cfg_attr(not(feature = "expensive-debug"), derive(Debug))]
```

Comment:
This non-`expensive-debug` derive and the explicit `expensive-debug` implementation below are the meat of this PR (all the rest is just moving things around to relax the requirements enabling usage of certain methods when only BorshSerialize or BorshDeserialize implemented and not both.